cmd/registry/scoring/parser.go

```diff
@@ -258,59 +295,147 @@ func validateBooleanThresholds(thresholds []*rpc.BooleanThreshold) []error {
+		if tp.Name.VersionID != "-" && tp.Name.VersionID != rName.Name.VersionID {
+			return fmt.Errorf("version mismatch in resource %q and target pattern %v", resourceInstance.ResourceName().String(), targetPattern)
 		}
-		if tp.Name.SpecID != "-" && tp.Name.SpecID != r.Name.SpecID {
-			return fmt.Errorf("spec mismatch in resource %s and target pattern %v", resourceName.String(), targetPattern)
+		if tp.Name.SpecID != "-" && tp.Name.SpecID != rName.Name.SpecID {
+			return fmt.Errorf("spec mismatch in resource %q and target pattern %v", resourceInstance.ResourceName().String(), targetPattern)
+		}
+
+		// Check if the filter matches with the resource
+		filterStr := targetPattern.GetFilter()
+		if filterStr != "" {
+			filter, err := filtering.NewFilter(filterStr, filtering.SpecFields)
+			if err != nil {
+				return fmt.Errorf("invalid filter %q: %s", filterStr, err)
+			}
+
+			rMap, err := filtering.SpecMapFromMessage(r.Spec)
+			if err != nil {
+				return fmt.Errorf("internal error while applying filter: %s", err)
+			}
+
+			match, err := filter.Matches(rMap)
+			if err != nil {
+				return fmt.Errorf("failed applying filter %q: %s", filterStr, err)
+			}
+
+			if !match {
+				return fmt.Errorf("target pattern filter %q does not match with resource %q", filterStr, resourceInstance.ResourceName().String())
+			}
 		}
 	case patterns.VersionName:
-		// Check if targetPattern and resourceName match in type
-		r, ok := resourceName.(patterns.VersionName)
+		// Check if targetPattern and resource match in type
+		r, ok := resourceInstance.(patterns.VersionResource)
 		if !ok {
-			return fmt.Errorf("resource %s doesn't match target pattern %s", r, tp)
+			return fmt.Errorf("resource %q doesn't match target pattern %q", r, tp)
+		}
+
+		// Convert ResourceName to VersionName
+		rName, ok := resourceInstance.ResourceName().(patterns.VersionName)
+		if !ok {
+			return fmt.Errorf("resourceName %q doesn't match target pattern %q", rName, tp)
 		}
 
 		// Check if the individual entities match
-		if tp.Name.ApiID != "-" && tp.Name.ApiID != r.Name.ApiID {
-			return fmt.Errorf("api mismatch in resource %s and target pattern %v", resourceName.String(), targetPattern)
+		if tp.Name.ApiID != "-" && tp.Name.ApiID != rName.Name.ApiID {
+			return fmt.Errorf("api mismatch in resource %s and target pattern %v", resourceInstance.ResourceName().String(), targetPattern)
 		}
-		if tp.Name.VersionID != "-" && tp.Name.VersionID != r.Name.VersionID {
-			return fmt.Errorf("version mismatch in resource %s and target pattern %v", resourceName.String(), targetPattern)
+		if tp.Name.VersionID != "-" && tp.Name.VersionID != rName.Name.VersionID {
+			return fmt.Errorf("version mismatch in resource %s and target pattern %v", resourceInstance.ResourceName().String(), targetPattern)
+		}
+
+		// Check if the filter matches with the resource
+		filterStr := targetPattern.GetFilter()
+		if filterStr != "" {
```

Comment:
Are these checks necessary? I recall building the filtering package to handle this case gracefully. This _should_ be a duplication of logic already handled by the package, and if it isn't, we should probably fix the filtering package to handle empty filters correctly.

Reply:
Done.

Verified that the filtering package does take care of empty filters.